### PR TITLE
[WIP] Pre pull images before running tests

### PR DESCRIPTION
**Pull request by aveshagarwal**
@ncdc 

fixes https://github.com/openshift/origin/issues/9444

### CODE REVIEW

File: hack/test-pre-pull-images.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+
+echo "[INFO] Prepulling container images"
+
+# https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/e2e-image-puller/e2e-image-puller.manifest
+# This list of images was created via:
+# grep -Iiroh "gcr.io/google_.*" "${OS_ROOT}" | \
+#     sed -e "s/[,\")}]//g" | awk '{print $1}' | sort | uniq | tr '\n' ' '
+
+for i in gcr.io/google_containers/busybox \
```
Comment (liggitt) [inline hack/test-pre-pull-images.sh]: would be nice if something would complain if changes to this list were needed...

Comment (ncdc) [inline hack/test-pre-pull-images.sh]: It'd also be nice to capture images used in the OpenShift tests as well

Comment (aveshagarwal) [inline hack/test-pre-pull-images.sh]: @liggitt that makes sense. Not sure about clean up just picked it up from that kube's manifest. 
@ncdc will update soon.

**Comment by aveshagarwal** [comment]
For openshift I have only searched in origin/test/ dir and could find only 2 unique images:
gcr.io/google_containers/pause, openshift/origin

Also if you have any suggestions to improve the image search method in this PR, let me know.

**Comment by marun** [comment]
I just saw a ci failure due to gcr.io temporarily not being resolveable:

https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/271/

In the case of the running select extended tests (in this case networking), the proposed approach of downloading all images would add unnecessary overhead where only a subset of images are required.  A better solution would be a pull-through caching registry, but upstream docker appears to be dragging their feet on allowing the registry to cache from private registries like gcr.io:

https://github.com/docker/distribution/issues/1431

**Comment by ncdc** [comment]
Our integrated registry supports pull-through, and I think it supports auth.

Given that most if not all of the e2e images are fairly static tags (i.e. not "latest"), we could consider pre-pulling and baking the images into the AMI.

**Comment by bparees** [comment]
bump.  this is still flaking regularly on us.

**Comment by ncdc** [comment]
@smarterclayton @bparees @stevekuznetsov @danmcp given what @marun wrote above, do we want to bake the e2e images into the amis?

**Comment by aveshagarwal** [comment]
@ncdc static tags will also change over time and given the number of images, wouldn't it require updating AMIs frequently?

**Comment by ncdc** [comment]
@aveshagarwal we update and rebuild the AMIs quite frequently

**Comment by ncdc** [comment]
We alternatively could pull the e2e images, retag them to 1 of our registries running in AWS, and then add code in the various e2e scripts to pull from our registry and then retag them back to gcr.io (this saves us from having to bake them in to the AMIs and it saves us from having to talk to gcr.io for each test)

**Comment by aveshagarwal** [comment]
Also thinking if baking so many (third party) images into AMIs would be good from security point of view.

**Comment by ncdc** [comment]
?

**Comment by aveshagarwal** [comment]
Just thinking out loud, If there is any security issue with any of the images, wouldn't keeping track and updating AMIs be an issue?

**Comment by ncdc** [comment]
These AMIs are strictly for running tests and as devenvs, and we rebuild them so often, the old ones get deleted soon enough.

**Comment by smarterclayton** [comment]
I don't see a problem with pre-pulling a few common images.

**Comment by ncdc** [comment]
The list has ~37 in it

File: hack/test-pre-pull-images.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+gcr.io/google_containers/ubuntu:14.04
+gcr.io/google_containers/update-demo:kitten
+gcr.io/google_containers/update-demo:nautilus
+gcr.io/google_containers/volume-ceph:0.1
+gcr.io/google_containers/volume-gluster:0.2
+gcr.io/google_containers/volume-iscsi:0.1
+gcr.io/google_containers/volume-nfs:0.6
+gcr.io/google_containers/volume-rbd:0.1
+gcr.io/google_samples/gb-redisslave:v1
+openshift/origin"
```
Comment (ncdc) [inline hack/test-pre-pull-images.sh]: don't pull origin; we build it

**Comment by danmcp** [comment]
@ncdc As long as individual tests associated with a particular pull aren't going to be dependent on particular versions of the images, it should be fine to cache them.

**Comment by ncdc** [comment]
@danmcp yeah these are pretty much all upstream tests in Kube. You can see the list of images in this PR

**Comment by kargakis** [comment]
Will we be able to use those images in extended tests?

**Comment by stevekuznetsov** [comment]
Poor Rosie is working overtime labeling this as needs-rebase. @aveshagarwal I'll close this one out for now, if you revisit this effort we can open it back up.

**Pull request closed without merging**